lib/attack.go

```diff
@@ -223,7 +223,61 @@ func Client(c *http.Client) func(*Attacker) {
 	return func(a *Attacker) { a.client = *c }
 }
 
-// A Rate of hits during an Attack.
+// A Rater defines the rate of hits during an Attack by
+// returning the duration an Attacker should wait until hiting
+// the next Target. When a negative value time is returned,
```

Comment:
(opinionated) `When a negative value time is returned` - I think Golang has multiple number of return values to adress cases like this, we can have something like `Wait(began, now time.Time) (value time.Duration, hasNext bool)`

Reply:
Having messed around with this some, I think the first alternative is actually better. The approach I've converged on is:

*   `Attack` should be responsible for terminating the attack after any non-zero duration has elapsed.
    *    This doesn't seem like the job of code that is supposed to be pacing hits.
*   `Attack` should track both the elapsed time of the attack and the number of hits sent.
    *    It should provide the `Pacer` (previously `Rater`) with both of these.
    *    This means the `Pacer` doesn't necessarily need to maintain internal state, so it's possible to preserve backwards compatibility with the current `Rate` struct.
    *    The Pacer is free to ignore these if its implementation doesn't need them.
*   `Pacer` still needs some way of stopping the attack that's not necessarily duration based.
    *   This allows for control logic to e.g. abort load tests when errors / latency are outside acceptable thresholds.

Thus, I've arrived at:

```go
type Pacer interface {
	Pace(elapsedTime time.Duration, elapsedHits uint64) (sleep time.Duration, stop bool)
}
```

Reply:
Yeah, I also think that `Pacer` shouldn't be responsible for `the elapsed time of the attack and the number of hits sent`. 

Also, in my opinion the name `Pacer` reflects the intention better than `Rater`.

Reply:
Agree with everything here. This is a great example of collective design! Happy about this outcome.

```
type Pacer interface {
	Pace(elapsedTime time.Duration, elapsedHits uint64) (sleep time.Duration, stop bool)
}
```